### ISSUE TITLE
EnvironmentConfiguration and charset problem

### ISSUE THREAD

**Issue by replay111** [jtwig/jtwig#334]
Hi,

I initialized Jtwig using Your example (```https://github.com/jtwig/jtwig-examples/blob/master/gradle-jtwig-spring-simple/src/main/java/org/jtwig/example/config/WebConfig.java```)

```java
@Bean
    public ViewResolver viewResolver () {
        JtwigViewResolver viewResolver = new JtwigViewResolver();
        viewResolver.setRenderer(JtwigRenderer.defaultRenderer());
        viewResolver.setPrefix("classpath:/templates/");
        viewResolver.setSuffix(".html.twig");
        return viewResolver;
    }
```

I wanted to disable strict mode so I was trying to use EnvironmentConfiguration like this:

```java
  @Bean
    public ViewResolver viewResolver() {

        EnvironmentConfiguration environmentConfiguration = EnvironmentConfigurationBuilder
                .configuration()
                    .render()
                        .withStrictMode(false)
                        .withOutputCharset(Charset.forName("UTF-8"))
                .and()
                    .parser()
                        .withoutTemplateCache()
                .and()
                .build();

        JtwigRenderer jtwigRenderer = new JtwigRenderer(environmentConfiguration);

        JtwigViewResolver viewResolver = new JtwigViewResolver();
        viewResolver.setRenderer(jtwigRenderer);

        viewResolver.setPrefix("classpath:/templates/");
        viewResolver.setSuffix(".html.twig");

        return viewResolver;
    }

```

but page in browser does not display my locale spacial chars. I am given:

```uÅ¼ytkownik``` instead of ```użytkownik```

But if I initialize viewResolver like this:

```
 @Bean
    public ViewResolver viewResolver() {

        JtwigViewResolver viewResolver = new JtwigViewResolver();

        viewResolver.setRenderer(JtwigRenderer.defaultRenderer());
        viewResolver.setPrefix("classpath:/templates/");
        viewResolver.setSuffix(".html.twig");
        viewResolver.setContentType("text/html; charset=UTF-8");

        return viewResolver;
    }

```

locales on my page are correct.

Can You tell me if I incorrectly used ```EnvironmentConfiguration``` to change default configuration? Or there is simpler way to toggle ```strictMode``` ?

Thx for help ;-)

Maciej.

**Comment by janoszen** [comment]
Although I'm not the author of jTwig, I believe your solution is the simplest one.

However, I had problems with UTF-8 too until I switched the underlying operating system to use UTF-8. I run Linux on Docker and I did these things in my Dockerfile:

```
RUN echo "en_US UTF-8" >/etc/locale.gen
RUN echo "en_US.UTF-8 UTF-8" >>/etc/locale.gen
RUN locale-gen

ENV LANG en_US.UTF-8
ENV LANGUAGE en_US:en
ENV LC_ALL en_US.UTF-8
```

**Comment by replay111** [comment]
@janoszen I am developing on OSX and with JTwig 3.1.1 I did not have such problems (with UTF-8). In jtwig 3.1.1 version there was simple way to toggle strictMode:

```java
@Bean
    public ViewResolver viewResolver() {
        JtwigViewResolver viewResolver = new JtwigViewResolver();
        viewResolver.setPrefix("classpath:/templates/");
        viewResolver.setSuffix(".twig");
        viewResolver.setCached(false);
	viewResolver.configuration().render().strictMode(true);
        return viewResolver;
    }
```

Thanks for Your example of configuration class - maybe I will use it as a pattern/sample to build my custom configuration. But I hope that there is simple way to change strictMode and set default encoding;-) In 3.1.1 it was out of the box - no special config required.

I want to be able to change strictMode in properties file (by using profiles dev/pod) as You mentioned to save my time with finding which variable in template is not initialized or initialized incorrectly.

I found also with new version of JTwig (5.x) that when I'm using Artffact:exploded in IDE, when I change something on template and update resources on Application Server (with IDE) those changes are not visible. I have to redeploy application or restart whole server. Again, with version 3.1.1 there was no problem, but... I prefer use new/current version than the older one ;)


And at the end I must confess that I am very vary newvie in Java - so if I am writing some stupid things... please let me know ;-)

**Comment by janoszen** [comment]
Welcome to the club, I've also switched to Java a couple of months ago. Luckily, I have a lot of people I can ask. One of the recommended me to set the env variables, or these runtime parameters to avoid character encoding problems:

```
-Dfile.encoding=utf-8 -Duser.language=en -Duser.country=US -Duser.variant=US
```

Anyway, if you're new to Java, you may want to try Jetty Embedded. Having switched from PHP and Hacklang, I find it much more pleasant than the whole servlet thing. (Especially deploying it into Amazon Container Service is a breeze.) Check my repo above for an example.

**Comment by replay111** [comment]
Hi,

I was using PHP (and sometimes still I'm using), but for a long time I wanted to learn Java. 
For me it is a very very slow process (lack of time unfortunately :( ). 
But lately I found some free time so I decided to refresh my knowledge and start using Java in web apps ;-). 
I am a admin/devops so I prefer using classic application servers like JBoss/Wildfly. It helps me with getting knowledge from developer area, but also from admin area ;-).

Back to my Issue... for now I have switched to Thymeleaf template engine.
It has great documentation and a lot of hints/solutions on StackOverflow. I was able to solve my "problem" with cacheing and encoding in a couple of minutes.
The main effort was in changing html templates. Luckily I did not have many templates to rewrite so it was not a such a very big problem. 

For now I will stay with Thymeleaf, but still will follow JTwig - which is great and @Lyncodev did great job  ;-)

**Comment by lyncodev** [comment]
@replay111 good luck :) Jtwig, however growing every month, is far from having a community like thymeleaf and some others, and obviously, makes complete sense for people with not a lot of experience with Java to adopt technologies where examples spring out of everywhere. Jtwig is powerful but still a young technology which needs growing and perfected, at the moment, I believe experienced Java devs are in better position to take the best out of it. Anyway, I'll try to shed some light on the encoding subject.

*How encoding works*

When it comes to encoding it's important to consider two things:

- Input Encoding
- Output Encoding

So, it's important that you let Jtwig know which input encoding you're expecting. If no encoding information is provided, then Jtwig will use the default JVM encoding `Charset.defaultCharset()`.

To get the input encoding you need to check your file editor (maybe IDE) which encoding it's being used on the template. That should be then specified to Jtwig. Let's assume your templates are being written in `ISO-8859-2`.

```java
EnvironmentConfigurationBuilder
                .configuration()
                    .resources()
                        .withDefaultInputCharset(Charset.forName("ISO-8859-2"))
                    .and()
                .build();
```

Jtwig can also be configured with output charset, that is, the encoding generated on the output. So, lets say for example you want to parse a template written using `ISO-8859-2` encoding and render it using `UTF-8` encoding, that can be achieved with the following:

```java
EnvironmentConfigurationBuilder
                .configuration()
                    .resources()
                        .withDefaultInputCharset(Charset.forName("ISO-8859-2"))
                    .and()
                    .render()
                        .withOutputCharset(Charset.forName("UTF-8"))
                    .and()
                .build();
```

Usually people struggle with encodings and end up by using the same encoding everywhere (for input, output and JVM), that way there is no need for specific Jtwig configuration.

In summary, provide Jtwig the encoding used to write the template and the desired output encoding and it will work like a charm.

**Comment by janoszen** [comment]
@lyncodev kind of a side note: I would really like to help because I want Jtwig to succeed. If you wanna point me at something I could help you with, even if it's just documentation, that'd be great.

**Comment by replay111** [comment]
OK, I've switched to Jtwig branch ;-) (great GIT) ;-)

In console I've checked my file encoding:

```login.html.twig: text/html; charset=utf-8```

So I setup input encoding for UTF-8... but still I am given some strange characters ;-)
I think that in wildfly I don't have change my encoding, because other sites with utf-8 looks ok...

Besides I've noticed that despite of having war:exploded, when I am making some changes in templates and tell server to update resources, those changes are still invisible. I have to restart whole appplication server... So it looks like option ```.parser().withoutTemplateCache()``` is ignored...

**Comment by lyncodev** [comment]
@replay111 that looks specific to your environment, I'm afraid I won't be able to help. @janoszen that would be very good. Since you have a lot of Twig experience I'm sure there is a lot to do. Especially on the documentation side of things.

**Comment by replay111** [comment]
Hm... version 3.1.1 works perfectly without any charset settings... 5.x works too when I am setting it like this: ``` viewResolver.setContentType("text/html; charset=UTF-8"); ``` but I am not able to disable cache in this way... 

Anyway thanx for any suggestions, I will try them and maybe find solution ;-)

**Comment by janoszen** [comment]
@replay111 if you're willing to share the code, I can take a look at it.
@lyncodev I think I'll start with writing a test suite of templates and their expected results as rendered in Twig. That should help us find bugs quickly.

**Comment by lyncodev** [comment]
@replay111 I give it a review too. @janoszen awesome!

**Comment by replay111** [comment]
OK, but probably during the weekend I will pull code to review ;-)
Many thanx for help!

**Comment by virtyvoid** [comment]
Greetings everybody, yesterday i decided to continue working on my old app with Spring & Jtwig.
It were using Jtwig 3.1.1 and so I upgraded it to latest release, 5.85.3.
I suffered whole last night with invalid encoding trying all solutions i could find.
At least 've tried: Servlet Filters, -Dfile.encoding=utf-8, manually configuring EnvironmentConfiguration with UTF8 in render/resources. Nothing at all.

My config:
```java	@Bean
	public JtwigViewResolver viewResolver() {
		JtwigViewResolver viewResolver = new JtwigViewResolver();
		viewResolver.setPrefix("/WEB-INF/pages/");
		viewResolver.setSuffix(".twig");
		viewResolver.setOrder(1);
		//viewResolver.setContentType("text/html; charset=utf8"); // why so?
		final EnvironmentConfiguration cfg = EnvironmentConfigurationBuilder.configuration()
				.render().withOutputCharset(StandardCharsets.UTF_8)
				.and().resources().withDefaultInputCharset(StandardCharsets.UTF_8).and().build();
		viewResolver.setRenderer(new JtwigRenderer(cfg));
		return viewResolver;
	}
```

At last i've tried to use `JtwigViewResolver.setContentType("text/html; charset=utf8");` and it did the job! Even with default environment configuration.

It's probably my question to the Developer @lyncodev , is this the only way to set correct encoding in Content-Type? Cuz even `render().withOutputCharset(StandardCharsets.UTF_8)` did nothing for me :(
For instance, with Thymeleaf same thing is clear and quite simple: `ThymeleafViewResolver.setCharacterEncoding("UTF-8")`. 
I mean i don't have to manually specify whole content type.

**Comment by janoszen** [comment]
@virtyvoid could you, by any chance, build a simple WAR file or give me a source repo so I can debug the issue for you? Also, which servlet container are you using?

**Comment by virtyvoid** [comment]
@janoszen Surely, i will do.
I am using Jetty 9.4.2.

**Comment by virtyvoid** [comment]
@janoszen Okay, i have uploaded simple project for you.
https://github.com/virtyvoid/TestWeb
That will be just awesome if you find a better solution for that problem.

p.s There also a little problem with accessing static methods through instance, i hadn't it in 3.1.1, but now it's bothering me. Probably you (or somebody else) have any clue on that.

**Comment by replay111** [comment]
Hi Guys!,

sorry for long silence, but I lately I haven't much time for coding.
Luckily today I had some spare time, so I pushed a piece of code with JTwig to gh repo.

@janoszen , @lyncodev if You still have some time and will, You can find my code in ```https://github.com/replay111/amareno/tree/jtwig``` (branch ```jtwig```) and review it.

If You have any suggestion (not only with JTwig) please let me know - probably	 I did many things against good practice of coding, so if there is something I can fix or do this better it will be great to know ;-)

**Comment by janoszen** [comment]
Hello everyone, I'll take a look shortly, but I'm really swamped at the moment.

**Comment by replay111** [comment]
@janoszen realx ;-) there is no hurry ;-) I am currently also involved in other tasks so currently don't have much time for mine own code... 
But if You find some ugly or to fix or to rewrite... write without hesitation :D

**Comment by janoszen** [comment]
@virtyvoid I took a look at your code, and the problem is not with jTwig, but with your `Content-Type` header. It returns as `text/html;charset=ISO-8859-1` which is not what you want for Cyrillic letters. Instead, it should return `text/html;charset=UTF-8`. I have read that spring framework allows a `produces` parameter for this purpose in `@RequestMapping` annotation, but I couldn't get it to work. I'm not a Spring user myself.

@replay111 can you check if this is the case with your code too please?

**Comment by virtyvoid** [comment]
@janoszen Sadly. But all worked fine in 3.1.1 though, anyway I appreciate your attempt to help me out. 
Will try to investigate by myself or move onto stackoverflow.
p.s Did you have an opportunity to check issue with static methods?

**Comment by janoszen** [comment]
@virtyvoid thanks, I'll try to debug it, maybe I can find out what part of the code sets the content type.

**Comment by virtyvoid** [comment]
@janoszen I did ran debugger and seems find out why 3.1.1 provides correct content-type.
Constructor of `JtwigResolver` has this:
`setContentType("text/html; charset=UTF-8");`
Latest version seemd to be missing that and that's why Spring uses his Default content type (cuz it's wasn't set through AbstractTemplateViewResolver).
I'd may be mistaken, but that's what i digged.

Well, i don't know, but maybe Jtwig should pass chosen render encoding to AbstractTemplateViewResolver's content type, or end-users should just specify everything on their own by calling `setContentType`.

**Comment by janoszen** [comment]
@lyncodev any input on the charset issue? Should jTwig set a content type or leave it to Spring to decide?

**Comment by replay111** [comment]
Hi,

I've changed annotation in login controller:

```java

    @RequestMapping(value = {"/login"}, produces = "text/html;charset=UTF-8")

```

but it did not change anything - page still does not shows polish characters...

@virtyvoid  so it looks like it is necessary to add:

```java
 viewResolver.setContentType("text/html; charset=UTF-8");
```

during the JTwig viewResolver initialisation (https://github.com/replay111/amareno/blob/jtwig/src/main/java/pl/mk/configuration/jtwig/JTwig.java)
